### PR TITLE
KZG Polynomial Commitment Scheme

### PR DESCRIPTION
**Pull request by thogiti**
I am working on the KZG commitment scheme. My goal is to write a self-containing wiki page for KZG so that beginners and intermediate skilled audience can read it and understand it. 

Why we need to learn KZG:
- Modern ZKSNARK = Functional Commitment Scheme + Compatible Interactive Oracle Proof (IOP)
- Ethereum Danksharding
- Other important Cryptography and Blockchain applications to preserve privacy

I will derive KZG by hands for a small finite field to visualize all the math happening behind KZG. This can understand KZG protocol flow.

https://github.com/thogiti/protocol-studies/blob/scaling/docs/wiki/Cryptography/KZG.md

### CODE REVIEW

**Review by taxmeifyoucan — COMMENTED** [review]
Thank you so much for creating this! It's big piece of content and I left few comments on the first part. I will need more time to dive into the actual crypto and also hoping someone else could provide a feedback

File: docs/wiki/Cryptography/KZG.md [hunk 1]
```diff
@@ -0,0 +1,528 @@
+# KZG Polynomial Commitment Scheme
+
+## [TLDR](#tldr)
+The KZG commitment scheme is like a cryptographic vault for securely locking away polynomials (mathematical equations) so that you can later prove you have them without giving away their secrets. It's like making a sealed promise that you can validate without ever having to open it up and show the contents. Using advanced math based on elliptic curves, it enables efficient, verifiable commitments that are a key part of making blockchain transactions more private and scalable. This scheme is especially important for Ethereum's upgrades, where it helps to verify transactions quickly and securely without compromising on privacy.
+
+The KZG Polynomial Commitment Scheme ensures that both commitments and evaluation proofs are of a fixed size, regardless of the polynomial's length, offering consistent and space-efficient cryptographic operations[^5][^6][^7].
+
+## [Motivation](#motivation)
+
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:9]: Great, thx.

File: docs/wiki/Cryptography/KZG.md [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+To make ZKSNARKs, you need to choose a PCS and IOP that work well together. The PCS is used to commit to a polynomial and prove the evaluation of the polynomial at a certain point without revealing the polynomial. The IOP is used to create a proof of knowledge, which is then turned into a non-interactive proof using a method called the Fiat-Shamir heuristic. This makes it possible to generate and verify the proof without needing to communicate in real-time, making ZKSNARKs useful for blockchain and distributed systems.
+
+### [Ethereum Danksharding](#ethereum-danksharding)
+KZG commitment scheme has emerged as a pivotal technology in the Ethereum ecosystem, particularly in the context of Proto-Danksharding and its anticipated evolution into Danksharding. This commitment scheme is a cornerstone of many Zero-Knowledge (ZK) related applications within Ethereum, enabling efficient and secure verification of data without revealing the underlying information.
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:22]: Great, thx. Shortened the ZKSNARKS explanation and expanded the KZG use cases in Ethereum based applications.

File: docs/wiki/Cryptography/KZG.md [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+
+Moreover, the KZG commitment scheme is not only limited to Ethereum but has broader applications in the blockchain and cryptographic communities. Its ability to commit to a polynomial and verify evaluations of the polynomial without revealing the polynomial itself makes it a versatile tool for various cryptographic protocols and applications. 
+
+
+## [Goal](#goal)
+Now that we are motivated to learn PCS, let us get started with defining what is our goal i.e. what is the exact problem we want to solve with KZG scheme. 
+
+Say we have a function or polynomial $f(x)$ defined as $f(x) = f_0 + f_1x + f_2x^2 + \ldots + f_dx^t$. The degree of $f(x)$ is $t$.
+
+Our main goal with KZG scheme is that we want to prove to someone that we know this polynomial without revealing the polynomial. What do we mean by without revealing by polynomial? We meant that without revealing the coefficients of the polynomial.
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:34]: Great, thx.

File: docs/wiki/Cryptography/KZG.md [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+
+Say we have a function or polynomial $f(x)$ defined as $f(x) = f_0 + f_1x + f_2x^2 + \ldots + f_dx^t$. The degree of $f(x)$ is $t$.
+
+Our main goal with KZG scheme is that we want to prove to someone that we know this polynomial without revealing the polynomial. What do we mean by without revealing by polynomial? We meant that without revealing the coefficients of the polynomial. 
+
+In practice what we exactly do is that we prove that we know a specific evaluation of this polynomial at a point $x=a$. 
+
+We write this, $f(a)$, for some $x=a$. 
+
+## [What we need to know before we discuss KZG](#what-we-need-to-know-before-we-discuss-kzg)
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:40]: Gtreat, thx.

File: docs/wiki/Cryptography/KZG.md [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+
+There are some important concepts we need to know before we can move further to understand KZG scheme. Fortunately, we can get an Engineering level understanding of the KZG scheme from just enough high school mathematics. We will try to gain some intuition on advanced concepts and their important properties without knowing them intimately. This can help us see the KZG protocol flow without bogged down by the advanced mathematics.
+
+We need to know:
+
+### [Modular Arithmetic](#modular-arithmetic)
+- Can you read a clock? Can you add/subtract two time values? In general can you do clock arithmetic? This is called Modular arithmetic. We only need to know how to add, subtract, multiply or divide numbers and apply modulus operation (the clock scale).
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:47]: Great, thx.

File: docs/wiki/Cryptography/KZG.md [hunk 6]
```diff
@@ -0,0 +1,528 @@
+# KZG Polynomial Commitment Scheme
+
+## [TLDR](#tldr)
+The KZG commitment scheme is like a cryptographic vault for securely locking away polynomials (mathematical equations) so that you can later prove you have them without giving away their secrets. It's like making a sealed promise that you can validate without ever having to open it up and show the contents. Using advanced math based on elliptic curves, it enables efficient, verifiable commitments that are a key part of making blockchain transactions more private and scalable. This scheme is especially important for Ethereum's upgrades, where it helps to verify transactions quickly and securely without compromising on privacy.
+
+The KZG Polynomial Commitment Scheme ensures that both commitments and evaluation proofs are of a fixed size, regardless of the polynomial's length, offering consistent and space-efficient cryptographic operations[^5][^6][^7].
+
+## [Motivation](#motivation)
+
+### [ZKSNARKs](#zksnarks)
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:10]: Great, thx. Shortened and wrote a brief paragraph for the Ethereum based KZG applications.

File: docs/wiki/Cryptography/KZG.md [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+
+## [KZG Protocol Flow](#kzg-protocol-flow)
+
+Let us reiterate on what is the problem we are solving with KZG protocol.
+
+We want prove that we know a specific evaluation of a function or polynomial at a point $x=a$ without revealing it.
+
+In the KZG commitment scheme, the roles of the Trusted Third Party, Prover, and Verifier are critical to its function and security. Here's how each contributes to the process:
+
+1. **Trusted Third Party (Setup Authority):** This entity is responsible for the initial setup phase of the KZG scheme. They generate the public parameters (PP) or Common Reference String (CRS) that will be used in the commitments and proofs, based on a secret that only they know. This secret is crucial for the construction of commitments but must be discarded (or kept extremely secure) after the setup to ensure the system's integrity. The trust in this party is fundamental because if the secret is mishandled or leaked, it could compromise the entire system.
```
Comment (taxmeifyoucan) [inline docs/wiki/Cryptography/KZG.md:247]: Might be worth also noting that this is an initial setup and the party is not later involved in proving-verifying

File: docs/wiki/Cryptography/KZG.md [hunk 8]
```diff
@@ -0,0 +1,537 @@
...
+
+Then, the trusted party sends the CRS to the Prover and the Verifier.
+
+In practice, this process is wrapped around a multi-party computation (MPC) where a secret is generated in such a way that, as long as at least one participant remains honest, the randomness of the secret can be guaranteed. 
+
+The trusted setup is a one-time procedure that generates a piece of data necessary for the cryptographic protocol to function. This data must be used every time the protocol is run, but once generated and the secrets are forgotten, no further participation from the creators of the ceremony is required. The trust in the setup comes from the fact that the secrets used to generate the data are securely discarded after the setup, ensuring that the data remains secure for future use
+
+Modern protocols often use a powers-of-tau setup, which involves hundreds of participants. The security of the final output depends on the honesty of at least one participant who does not publish their secret. This approach is considered "close enough to trustless" in practice, making it a practical solution for cryptographic protocols that require a trusted setup.
```
Comment (taxmeifyoucan) [inline docs/wiki/Cryptography/KZG.md:290]:
```suggestion
Modern protocols often use a powers-of-tau setup, which involves even thousands of participants. The security of the final output depends on the honesty of at least one participant who does not publish their secret. This approach is considered "close enough to trustless" in practice, making it a practical solution for cryptographic protocols that require a trusted setup. 
```

File: docs/wiki/Cryptography/KZG.md [hunk 9]
```diff
@@ -0,0 +1,537 @@
...
+
+
+### [Finite Field of order prime p](#finite-field-of-order-prime)
+
+A finite field of order prime $p$, we denote it by $\mathbb F_p$, is a special set of numbers, { $\{1, 2, 3, \ldots, p-1\}$ },  where you can do all the usual math operations (addition, subtraction, multiplication, and division, except by zero) and still follow the rules of arithmetic. 
+
+The "order" of this set is the number of elements it contains, and for a finite field of order prime $p$, this number is a prime number. The most common way to create a $\mathbb F_p$ is by taking the set of all integers greathan or equal to $0$ and dividing them by $p$, keeping only the remainders. This gives us a set of numbers from $0$ to $p-1$ that can be used for arithmetic operations. For example, if $p = 5$, the set would be {0, 1, 2, 3, 4}, and you can add, subtract, multiply, and divide these numbers in a way that follows the rules of arithmetic. This set is a finite field of order 5, we denote this by $\mathbb F_5$, because it has exactly 5 elements, and it's a prime number.
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:60]: Great, thx.

File: docs/wiki/Cryptography/KZG.md [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+
+In genral, when we define a finite field, we define, the order $p$ of the field and an arithemetic operation like addition or multiplication. If it is addition, we denote the field by $(\mathbb F_p, +)$. If it is multiplication, we denote it by $(\mathbb F^*_p, +)$. The `*` is telling us to exclude the zero element from our field so that we can satisfy all the required properties of the finite field i.e. mainly we can divide the numbers and find inverse of all elements. If we include the zero element, we can't find the inverse of zero element.
+
+### [Group](#group)
+A Group is a similar concept as finite field with some small changes. In a Group, we only have one arithmetic operation on the set, usually addition or multiplication as opposed to in a finite field we have addition and multiplication both and this group must satisfy few properties which we will discuss below with an example. We denote a Group by ($\mathbb G, +)$ for a Group with addition as the group operation, ($\mathbb G^*, .)$ for Group with multiplication operation; the `*` is telling to exclude zero element to avoid division by zero.
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:67]: Great, thx.

File: docs/wiki/Cryptography/KZG.md [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+  - $3^4 = 81 \equiv 4 \mod 7$
+  - $3^5 = 243 \equiv 5 \mod 7$
+  - $3^6 = 729 \equiv 1 \mod 7$ (and since we've reached the identity after hitting all elements, $3$ is a generator)
+
+You can verify that $5$ is also a generator for our multiplicative group ($\mathbb G^*_7, .)$ modulo $7$. 
+
+In the next section, we will learn how generators enable the KZG commitment scheme to function as an efficient, secure, and verifiable method of committing to polynomials, making it a powerful tool for cryptographic protocols, particularly in blockchain technologies where these qualities are very important.
+
+
+### [Why Primes for Modulo Operations in Fields](#why-primes-for-modulo-operations-in-fields)
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:134]: Great, thx.

File: docs/wiki/Cryptography/KZG.md [hunk 12]
```diff
@@ -0,0 +1,537 @@
...
+
+So, in essence, the Strong Diffie-Hellman assumption tells us that just knowing the outcomes of individual secrets (recipes) doesn't help anyone crack the result of combining those secrets. This is a cornerstone of secure communication, ensuring that even if someone knows the separate pieces, the combined secret remains safe and unguessable.
+
+
+### [Pairing Function](#pairing-function)
+Say we have a generator $g$ in the group $\mathbb G^\*_p$ and $a, b$ are any elements in the finite field $\mathbb F^*_p$ and $g^a$, $g^b$ are some elements in the group $\mathbb G^\*_p$. 
+
+A pairing function is a special kind of function that takes two inputs and produces a single output with two important properties, biliner and non-degenrate.
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:189]: Great, thx.

File: docs/wiki/Cryptography/KZG.md (hunk 8 above)
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:290]: Great, thx.

File: docs/wiki/Cryptography/KZG.md (hunk 7 above)
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:247]: Great, thx.

File: docs/wiki/Cryptography/KZG.md [hunk 13]
```diff
@@ -0,0 +1,537 @@
...
+
+2. **Simplicity of Construction:** When the modulus is a prime number, the field's construction is straightforward. The elements of the field are simply the set of integers  { ${1, 2, 3, \ldots, p-1}$ }, and the field operations (addition, subtraction, multiplication, and division) are performed modulo $p$. For non-prime moduli, constructing a field requires more complex structures, such as polynomial rings.
+
+3. **Guaranteed Field Properties:** The use of a prime modulus guarantees the satisfaction of field properties. A finite field (also known as a Galois field) is a field that contains a finite number of elements. For a set to be a field, it must satisfy several properties, including the existence of additive and multiplicative identities, the existence of additive and multiplicative inverses for every element, and the commutative, associative, and distributive laws for addition and multiplication. A prime modulus ensures all these properties are met.
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:142]: Great, thx.

File: docs/wiki/Cryptography/KZG.md [hunk 14]
```diff
@@ -0,0 +1,537 @@
...
+- **PSE's Summa, Proof of Solvenecy Protocol**: Ethereum Foundation's PSE group project, Summa, leverages KZG commitments in its Proof of Solvency protocol. This allows centralized exchanges and custodians to demonstrate that their total assets exceed their liabilities, all while keeping user balance information private.
+  
+- **Scroll's zkRollups**: Scroll, a native zkEVM Layer 2 for Ethereum, uses KZG to generate commitments to a collection of polynomials that encapsulate computations. This allows verifiers to request evaluations at random points to validate the accuracy of the computation represented by the polynomials.
+
+- **Jellyfish**: Jellyfish employs the KZG commitment scheme to generate commitments to polynomials during the commitment phase. It leverages the homomorphic properties of KZG for efficient evaluation of the polynomial at any point without revealing its coefficients.
+
+- **Hyperplonk**: Hyperplonk utilizes the multi-linear KZG commitment, indicating its application in scenarios requiring multi-linear polynomial commitments.
+
+
+## [Goal](#goal)
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:35]: Apologies. I am unable to see any headers with extra space. All headers are formatted the same way. 
Can you please point me to an example in the article? Thank you.

File: docs/wiki/Cryptography/KZG.md [hunk 15]
```diff
@@ -0,0 +1,537 @@
...
+
+When we do modular arithmetic operations in the finite field $\mathbb F_p$, we have a nice "wrap around" property i.e. the field behaves as if it "wraps around" after reaching $(p - 1)$. 
+
+In genral, when we define a finite field, we define, the order $p$ of the field and an arithemetic operation like addition or multiplication. If it is addition, we denote the field by $(\mathbb F_p, +)$. If it is multiplication, we denote it by $(\mathbb F^*_p, +)$. The `*` is telling us to exclude the zero element from our field so that we can satisfy all the required properties of the finite field i.e. mainly we can divide the numbers and find inverse of all elements. If we include the zero element, we can't find the inverse of zero element.
```
[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:64]: Great, thx.

File: docs/wiki/Cryptography/KZG.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (taxmeifyoucan) [reply docs/wiki/Cryptography/KZG.md:35]: Sorry I haven't been clear, I meant a new line under the header. For example headers on line 52 and 56, the former starts the paragraph without a new line and latter adds it 
image

[the remark this answers is not in the document]
Comment (thogiti) [reply docs/wiki/Cryptography/KZG.md:35]: Great, thx. Fixed.

**Comment by thogiti** [comment]
I've added a KZG Batch mode section in the commit hash [a196800](https://github.com/eth-protocol-fellows/protocol-studies/pull/118/commits/a196800304c4089e4a8c65d4b9783ecbe478b5b4).

**Comment by thogiti** [comment]
@taxmeifyoucan KZG wiki page is complete. I merged it into scaling branch.

**Comment by taxmeifyoucan** [comment]
Great, thanks! I will give it another read before merging to main

**Pull request merged**